django/contrib/admin/options.py

```diff
@@ -800,6 +800,12 @@ def get_actions(self, request):
             for func, name, desc in actions
         )
 
+        # Hide actions without permissions
+        model_perms = self.get_model_perms(request)
+        if not model_perms['delete']:
```

Comment:
I don't see the advantage of going through `get_model_perms()`. How about: 
`if 'delete_selected' in actions and not self.has_delete_permission(request):`